client/cli/src/grpc_client.rs

```diff
@@ -53,80 +53,135 @@ pub struct GRPCClient {
+    pub fn new(
+        host: &str,
+        ac_port: u16,
+        json_rpc_port: u16,
+        waypoint: Option<Waypoint>,
+    ) -> Result<Self> {
+        let client = AdmissionControlClientBlocking::new(host, ac_port);
         // If waypoint is present, use it for initial verification, otherwise the initial
         // verification is essentially empty.
         let initial_trusted_state = match waypoint {
             Some(waypoint) => TrustedState::from_waypoint(waypoint),
             None => TrustedState::new_trust_any_genesis_WARNING_UNSAFE(),
         };
+        let json_rpc_client = JsonRpcClientBlocking::new(host, json_rpc_port);
         Ok(GRPCClient {
             client,
+            json_rpc_client,
             trusted_state: initial_trusted_state,
             latest_epoch_change_li: None,
         })
     }
+    /// Sends JSON request `request`, performs basic checks on the payload, and returns Ok(`result`),
+    /// where `result` is the payload under the key "result" in the JSON RPC response
+    /// If there is an error payload in the JSON RPC response, throw an Err with message describing the error
+    /// payload
+    fn send_json_request(&mut self, request: serde_json::Value) -> Result<serde_json::Value> {
+        let response = self
+            .json_rpc_client
+            .send_with_retry(request)?
+            .error_for_status()
+            .map_err(|e| format_err!("Server returned error: {:?}", e))?;
+
+        // check payload
+        let data: serde_json::Value = response.json()?;
+
+        // check JSON RPC protocol
+        let json_rpc_protocol = data.get("jsonrpc");
+        ensure!(
+            json_rpc_protocol == Some(&serde_json::Value::String("2.0".to_string())),
+            "JSON RPC response with incorrect protocol: {:?}",
+            json_rpc_protocol
+        );
+
+        if let Some(error) = data.get("error") {
+            bail!("Error in JSON RPC response: {:?}", error);
+        }
+
+        if let Some(result) = data.get("result") {
+            Ok(result.clone())
+        } else {
+            bail!("Received JSON RPC response with no result payload");
+        }
+    }
 
     /// Submits a transaction and bumps the sequence number for the sender, pass in `None` for
     /// sender_account if sender's address is not managed by the client.
     pub fn submit_transaction(
         &mut self,
         sender_account_opt: Option<&mut AccountData>,
-        req: &SubmitTransactionRequest,
+        transaction: SignedTransaction,
     ) -> Result<()> {
-        let mut resp = self
-            .client
-            .submit_transaction(req.clone())
-            .map_err(Into::into);
+        // form request
+        let payload = hex::encode(lcs::to_bytes(&transaction).unwrap());
+        let request = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "submit",
+            "params": [payload],
+            "id": "Libra CLI client"
```

Comment:
the ID here should not be a fixed string, maybe an uuid or an random number would be better . help detect possible failure on the server side to conform to protocol (returning same id)

Reply:
id is more client concept than server side in this protocol. Server can't expect client to cooperate and send random ids. Id is to help clients structure work (e.g. in batching calls)
that being said, can we just use integers instead of string (similar to tests) for consistency